Add multi-process guide to docs. (#6796)
jax-ml/jax: docs/multi_process.md

# Using JAX in multi-host and multi-process environments

(test-ref)=
## Introduction

> This guide explains how to use JAX in environments such as Cloud TPU pods where

Reviewer: Suggestion @skye : Maybe (but only maybe) - add a link for those users who aren't aware of what the pods are:

```suggestion
This guide explains how to use JAX in environments such as [Cloud TPU](https://cloud.google.com/tpu) pods where
```

Author: Great idea!

accelerators are spread across multiple CPU hosts or JAX processes. We’ll refer
to these as “multi-process” environments.

This guide specifically focuses on how to use collective communication
operations (e.g. {func}`jax.lax.psum`) in multi-process settings, although other
communication methods may be useful too depending on your use case (e.g. RPC,
[mpi4jax](https://github.com/mpi4jax/mpi4jax)). If you’re not already familiar
with JAX’s collective operations, we recommend starting with the

> {doc}`/jax-101/06-parallelism` notebook. Note that an important feature of

Reviewer: Quick question/suggestion @skye : would the name of this page render on RTD when you use {doc}`/link`? If not, maybe it would help some users if there's a name of the notebook - _Parallel Evaluation in JAX_ (It's the 6th one but users may not see the numerical order when they come across JAX-101 at https://jax.readthedocs.io/en/latest/jax-101/index.html).

For example:

```suggestion
{doc}`/jax-101/06-parallelism` notebook (Parallel Evaluation in JAX). Note that an important feature of
```

Author: It does render the title! I forgot to post the link to the rendered version: https://jax--6796.org.readthedocs.build/en/6796/multi_process.html

(In general, you can see the rendered version of docs PRs by clicking on the "Details" of the "docs/readthedocs.org:jax" check)

multi-process environments is direct communication links between accelerators,

> e.g. the high-speed interconnects for Cloud TPUs or NCCL for GPUs. This is what

Reviewer: Suggestion @skye :

- Maybe some users may not be aware of what NCCL stands for. Perhaps you could spell out the acronym like `NVIDIA Collective Communication Library [thing] (NCCL) for GPUs` 🤷  (that's what NCCL stands for, right? 😃 )

Reviewer: The whole doc is amazing and well-written. I'm learning a lot here 👍 Thanks @skye

Author: As a compromise, I added a link to https://developer.nvidia.com/nccl where they define the acronym, because the expanded acronym is painfully long :P

Thanks, glad you're finding it useful :D

accelerators.

## Multi-process programming model

Key concepts:
*   All JAX processes must be manually run.
*   Each process has a distinct set of _local_ devices it can address. The
    _global_ devices are the set of all devices across all processes.
*   Use standard JAX parallelism APIs like {func}`~jax.pmap` and
    {func}`~jax.experimental.maps.xmap`. Each process “sees” _local_ input and
    output to parallelized functions, but communication inside the computations
    is _global_.
*   Make sure all processes run the same parallel computations in the same
    order.

### Launching JAX processes

Unlike other distributed systems where a single controller node manages many
worker nodes, JAX uses a “multi-controller” programming model where each JAX
Python process runs independently, sometimes referred to as an [Single Program,
Multiple Data (SPMD)](https://en.wikipedia.org/wiki/SPMD) model. Generally, the
same JAX Python program is run in each process, with only slight differences
between each process’s execution (e.g. different processes will load different
input data). Furthermore, **you must manually run your JAX program on each

> host!** I.e., JAX doesn’t automatically start multiple processes from a single

Reviewer: Suggestion @skye : if it's a new sentence, then maybe

```suggestion
host!** That is, JAX doesn’t automatically start multiple processes from a single
```

🤷

Author: I just got rid of the I.e.

Reviewer: IE = internet explorer

computations on the 8 TPU cores attached directly to that host (see the [Cloud
TPU System Architecture](https://cloud.google.com/tpu/docs/system-architecture)
documentation for more details). You can see a process’s local devices via
{func}`jax.local_devices()`.

**The _global_ devices are the devices across all processes.** A computation can
span devices across processes and perform collective operations via the direct
communication links between devices, as long as each process launches the
computation on its local devices. You can see all available global devices via
{func}`jax.devices()`. Note that a process’s local devices are always a subset
of the global devices.

### Running multi-process computations

So how do you actually run a computation involving cross-process communication?
**Use the same parallel evaluation APIs that you would in a single process!**

For example, {func}`~jax.pmap` can be used to run a parallel computation across
multiple processes. (If you’re not already familiar with how to use
{func}`~jax.pmap` to run across multiple devices within a single process, check
out the {doc}`/jax-101/06-parallelism` notebook.) Each process should call the

> same pmapped function and pass in arguments to be mapped across its _local_

Reviewer: Question (newbie) @skye : is "pmapped" the same "jitted" - part of the common JAX lingo? If so, then ignore the following suggestion: "`pmap`ped function" 😄 

```suggestion
same `pmap`ped function and pass in arguments to be mapped across its _local_
```

Author: Hm `pmap`ped is probably more accurate, but I'm gonna claim pmapped as official lingo because it looks much nicer :)

Reviewer: @skye It's also cooler! pmapped 🚀 xmapped 🚀 (@saeta)

# Using JAX in multi-host and multi-process environments

> (test-ref)=

Reviewer 2: remove?

Author: Oops, good catch

{doc}`/jax-101/06-parallelism` notebook. Note that an important feature of
multi-process environments is direct communication links between accelerators,
e.g. the high-speed interconnects for Cloud TPUs or
[NCCL](https://developer.nvidia.com/nccl) for GPUs. This is what allows
collective operations to run across multiple process’ worth of accelerators.

## Multi-process programming model

Key concepts:
*   All JAX processes must be manually run.
*   Each process has a distinct set of _local_ devices it can address. The
    _global_ devices are the set of all devices across all processes.
*   Use standard JAX parallelism APIs like {func}`~jax.pmap` and
    {func}`~jax.experimental.maps.xmap`. Each process “sees” _local_ input and
    output to parallelized functions, but communication inside the computations
    is _global_.
*   Make sure all processes run the same parallel computations in the same
    order.

### Launching JAX processes

Unlike other distributed systems where a single controller node manages many
worker nodes, JAX uses a “multi-controller” programming model where each JAX
Python process runs independently, sometimes referred to as an [Single Program,

> Multiple Data (SPMD)](https://en.wikipedia.org/wiki/SPMD) model. Generally, the

Reviewer 2: Since "SPMD" is in the glossary, you can use ``` {term}`SPMD` ``` to link there: https://jax.readthedocs.io/en/latest/glossary.html#term-SPMD

There might be some other terms in this guide as well that could be linked or added there

Author: Didn't know this, cool! I just changed this one for now.
